lib/ansible/modules/windows/win_mssql_query.ps1

```diff
@@ -0,0 +1,72 @@
+#!powershell
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2018, Daniele Lazzari  <lazzari@mailup.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#Requires -Module Ansible.ModuleUtils.Legacy
+
+$ErrorActionPreference = "Stop"
+$result = @{
+    'changed' = $False;
+    'output'  = @()
+}
+Function Invoke-Query {
+    Param(
+        [string]$Query,
+        [int]$QueryTimeout,
+        [string]$Database,
+        [string]$ServerInstance,
+        [string]$ServerInstanceUser,
+        [Security.SecureString]$ServerInstancePassword,
+        [bool]$CheckMode
+    )
+    $output = @()
+    try {
+        if (!($CheckMode)) {
+            if (!($ServerInstanceUser) -or !($ServerInstancePassword)) {
+                $res = Invoke-SqlCmd -ServerInstance $ServerInstance -Database $Database -Query $Query -QueryTimeout $QueryTimeout
+            }
+            else {
+                $SQLCredentials = New-Object -TypeName System.Management.Automation.PSCredential ($ServerInstanceUser, $ServerInstancePassword)
+                $res = Invoke-SqlCmd -ServerInstance $ServerInstance -Query $Query -Database $Database -Credential $SQLCredentials -QueryTimeout $QueryTimeout
+            }
+        }
+    }
+    catch {
+        $result.query = $Query
+        $result.database = $Database
+        Fail-Json -obj $result -message "Error: $($_.exception.message)"
+    }
+    if ($res) {
+        Switch ($res.GetType()|Select-Object -ExpandProperty Name) {
```

Comment:
You might be better off doing `if ($res -is [Array])` and handle that accordingly. Having codepaths based on the string of the type seems wrong to me.

Reply:
I don't like it too, but the problem here is that `if ($res -is [Array])` returns true whether is an Object[] or a DataRow.

Reply:
That's interesting [DataRow](https://docs.microsoft.com/en-us/dotnet/api/system.data.datarow?view=netframework-4.7.2) doesn't show any indication that it is an Array. You could also try `$res.GetType().IsArray` and iterate from there.

Should we be only getting the first entry if the value is `Object[]`? It seems like if it is an array we would be returning all `DataRows` in that array, e.g. make `$result.output` a list of `DataRow` values.

Reply:
Sorry, I was wrong :(
```
$res = Invoke-Sqlcmd -ServerInstance mysql -Query "Select top 2 name from sys.databases"
Is Array: $($res -is [Array])
Is Array: True
$res = Invoke-Sqlcmd -ServerInstance mysql -Query "Select top 1 name from sys.databases"
Is Array: $($res -is [Array])
Is Array: False
```
So I think that the code above could become:
```
if ($res) {
    if ($res -is [Array]){
        $Columns = $res[0].Table.Columns.Caption
    } 
    else {
        $Columns = $res.Table.Columns.Caption
    }
    $output = @($res |Select-Object -Property $Columns)
}
```